unit/java_bytecode/java_bytecode_parse_generics/parse_generic_class.cpp

```diff
@@ -0,0 +1,274 @@
+      {
+        const symbolt &class_symbol=new_symbol_table.lookup("java::generics$element");
+        const typet &symbol_type=class_symbol.type;
+
+        REQUIRE(symbol_type.id()==ID_struct);
+        class_typet class_type=to_class_type(symbol_type);
+        REQUIRE(class_type.is_class());
+        java_class_typet java_class_type=to_java_class_type(class_type);
+        REQUIRE(is_java_generics_class_type(java_class_type));
+        java_generics_class_typet java_generics_class_type=
+          to_java_generics_class_type(java_class_type);
+
+        const struct_union_typet::componentt &elem=
+          java_generics_class_type.get_component("elem");
+        const typet &elem_type=java_class_type.component_type("elem");
+
+        REQUIRE(is_java_generic_parameter(elem_type));
+      }
+    }
+  }
+  GIVEN("Some class files with generic type variable")
+  {
+    WHEN("Parsing a class with type variable")
+    {
+      java_lang->parse(java_code_stream, "generics$element.class");
+
+      symbol_tablet new_symbol_table;
+      java_lang->typecheck(new_symbol_table, "");
+
+      java_lang->final(new_symbol_table);
+
+      REQUIRE(new_symbol_table.has_symbol("java::generics$element"));
+      THEN("The symbol type should be generic")
+      {
+        const symbolt &class_symbol=new_symbol_table.lookup("java::generics$element");
+        const typet &symbol_type=class_symbol.type;
+
+        class_typet class_type=to_class_type(symbol_type);
+        java_class_typet java_class_type=to_java_class_type(class_type);
+        java_generics_class_typet java_generics_class_type=
+          to_java_generics_class_type(java_class_type);
+
+        REQUIRE(java_generics_class_type.generic_types().size()==1);
+        typet &type_var=java_generics_class_type.generic_types().front();
+        REQUIRE(is_java_generic_parameter(type_var));
+        java_generic_parametert generic_type_var=to_java_generic_parameter(type_var);
+        REQUIRE(
+          generic_type_var.type_variable().get_identifier()==
+          "java::generics$element::E");
+        typet &sub_type=generic_type_var.subtype();
+        REQUIRE(sub_type.id()==ID_symbol);
+        symbol_typet &bound_type=to_symbol_type(sub_type);
+        REQUIRE(bound_type.get_identifier()=="java::java.lang.Object");
+     }
+    }
+  }
+
+  GIVEN("Some class files with generic type variable")
+  {
+    WHEN("Parsing a class with bounded type variable")
+    {
+      java_lang->parse(java_code_stream, "generics$bound_element.class");
+
+      symbol_tablet new_symbol_table;
+      java_lang->typecheck(new_symbol_table, "");
+
+      java_lang->final(new_symbol_table);
+
+      REQUIRE(new_symbol_table.has_symbol("java::generics$bound_element"));
+      THEN("The symbol type should be generic")
+      {
+        const symbolt &class_symbol=new_symbol_table.lookup("java::generics$bound_element");
+        const typet &symbol_type=class_symbol.type;
+
+        REQUIRE(symbol_type.id()==ID_struct);
```

Comment:
For clarity, you might want to pull these checks into a single function:

```
java_class_typet CheckIsJavaClass(const typet &type)
{
  REQUIRE(symbol_type.id()==ID_struct);
  class_typet class_type=to_class_type(symbol_type);
  REQUIRE(class_type.is_class());
  return to_java_class_type(class_type);
}
```

I've done something like this in an unrelated PR: [write_expr.cpp](https://github.com/thk123/cbmc/blob/6120ca235df84bfcf96167c8412ea2b08a3e76a2/unit/src/expr/require_expr.cpp) which you might consider using as a model. I think it helps keep the intent of the tests clear and reduces total number of lines in the test.